types-tests/message.test-d.ts

```diff
@@ -1,14 +1,41 @@
-import { expectType } from 'tsd';
-import { App, MessageEvent, BotMessageEvent } from '../';
+import { expectNotType, expectType } from 'tsd';
+import { App, MessageEvent, BotMessageEvent, MessageRepliedEvent, MeMessageEvent, MessageDeletedEvent, ThreadBroadcastMessageEvent, MessageChangedEvent, EKMAccessDeniedMessageEvent } from '..';
 
 const app = new App({ token: 'TOKEN', signingSecret: 'Signing Secret' });
 
 expectType<void>(
+  // TODO: Resolve the event type when having subtype in a listener constraint
+  // app.message({pattern: 'foo', subtype: 'message_replied'}, async ({ message }) => {});
```

Comment:
We haven't agreed to implement this yet but this could be beneficial for some use cases. There are  a few more things to consider:

* Should we enable developers to pass multiple subtypes here?
  * If yes, the `message` object type in the scenario is not fully deterministic (it can be a bit more specific than `MessageEvent` but it's still a union type).
  * If no, for handling multiple types of subtyped message events, the right way is to have different event listeners for each subtype (some of internal code can be reused).
* Is `pattern` a good key name for the pattern here?
* `app.message<"message_replied">` vs `app.message(constraints` or should we support both?
  * `app.action` and `app.options` are not consistent at this point (refer to my comments below)

Reply:
I like the idea of having a constraints object here.  `subtype` in the constraints object makes sense. `pattern` seems like a reasonable key name. 

If we wanted to support multiple subtypes in one listener, what would that look like? An array for `subtype: ['message_replied']`. Having an array in a constraint object would be new, but if using one listener for multiple subtypes is a usecase we think is common and should handle, it makes sense. Maybe @shaydewael can chime in on this.

My gut feeling is `app.message(constraints` is more important. Does `app.options` not support `type` in constraints?

Reply:
We have another pull request making changes to the `app.message` listener signature: https://github.com/slackapi/bolt-js/pull/761

>one more suggestion for this PR: can we add some type tests within type-tests/message.test-d.ts that illustrate (and validate) the expected usage? 

Although Ankur asked the contributor to add more tests, we can handle the tests for the developer. Should we merge this PR before starting to work on the change we're discussing here?

@stevengill 
>My gut feeling is app.message(constraints is more important. Does app.options not support type in constraints?

As I mentioned the added tests for `app.options` in this PR, it just does not work. I think we should make the developer experience consistent across the listeners but currently there are some gaps. Refer to the TODO/FIXME comments in this PR for details.